### PR TITLE
[SPARK-33901][SQL] Fix Char and Varchar display error after DDLs

### PR DESCRIPTION
**Pull request by yaooqinn**
### What changes were proposed in this pull request?

After CTAS / CREATE TABLE LIKE / CVAS/ alter table add columns, the target tables will display string instead of char/varchar

### Why are the changes needed?

bugfix

### Does this PR introduce _any_ user-facing change?

no
### How was this patch tested?

new tests

### CODE REVIEW

**Comment by yaooqinn** [comment]
@cloud-fan @maropu @HyukjinKwon thanks

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/CharVarcharUtils.scala [hunk 1]
```diff
@@ -126,6 +126,16 @@ object CharVarcharUtils extends Logging {
     }
   }
 
+  /**
+   * Re-construct the original schema from the type string in the given metadata of each field.
+   */
+  def restoreCharVarCharInSchema(schema: StructType): StructType = {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/CharVarcharUtils.scala:132]: there is a `getRawType`, shall we name it `getRawSchema`?

Comment (yaooqinn) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/CharVarcharUtils.scala:132]: sgtm

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveStrategies.scala [hunk 2]
```diff
@@ -226,7 +227,8 @@ case class RelationConversions(
             tableDesc.partitionColumnNames.isEmpty && isConvertible(tableDesc) &&
             SQLConf.get.getConf(HiveUtils.CONVERT_METASTORE_CTAS) =>
         // validation is required to be done here before relation conversion.
-        DDLUtils.checkDataColNames(tableDesc.copy(schema = query.schema))
+        val tableSchema = CharVarcharUtils.replaceCharVarcharWithStringInSchema(query.schema)
+        DDLUtils.checkDataColNames(tableDesc.copy(schema = tableSchema))
```
Comment (cloud-fan) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveStrategies.scala:231]: do we really need to restore char/varchar for this `checkDataColNames`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala [hunk 3]
```diff
@@ -705,7 +705,7 @@ class SessionCatalog(
             schema = plan.output.toStructType)
       }.getOrElse(throw new NoSuchTableException(globalTempViewManager.database, table))
     } else {
-      getTableMetadata(name)
+      getTableRawMetadata(name)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala:708]: why do we need this change?

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveStrategies.scala (hunk 2 above)
Comment (yaooqinn) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveStrategies.scala:231]: Oh , my mistake

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala (hunk 3 above)
Comment (yaooqinn) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala:708]: AlterTableAddColumnsCommand will mess up the original schema or I just get the raw schema there？

Comment (yaooqinn) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala:708]: Ideally, this method should return the raw type of a table

Comment (yaooqinn) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala:708]: It is also related to the API `spark.catalog.listColumns` which returns a list of columns for the given table/view

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala:708]: It's a bit weird that `getTempViewOrPermanentTableMetadata` and `getTableMetadata` is not consistent. How many caller sides we need to update in this case?

Comment (yaooqinn) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala:708]: I have checked all the callers,` spark.catalog.listColumns` and AlterTableAddColumnsCommand only will directly reference the schema here.

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala:708]: if it's only 2 places, let's update the caller sides.

Comment (yaooqinn) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala:708]: Ok.
Just a little wired that just change the data type back and forth in the caller

File: sql/core/src/test/resources/sql-tests/inputs/charvarchar.sql [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+show create table char_tbl2;
+desc formatted char_tbl2;
+desc formatted char_tbl2 c;
+
+create table char_tbl3 like char_tbl;
+desc formatted char_tbl3;
+desc formatted char_tbl3 c;
+show create table char_tbl3;
+
+create view char_view as select * from char_tbl;
```
Comment (cloud-fan) [inline sql/core/src/test/resources/sql-tests/inputs/charvarchar.sql:16]: can we test alter view as well?

**Review by cloud-fan — APPROVED** [review]

File: sql/core/src/test/resources/sql-tests/inputs/charvarchar.sql (hunk 4 above)
Comment (yaooqinn) [reply sql/core/src/test/resources/sql-tests/inputs/charvarchar.sql:16]: More tests were added. Looks like more DDLs will mess up the table schema

**Comment by yaooqinn** [comment]
retest this please

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master/3.1!

**Pull request closed without merging**